lib/ProducerImpl.cc

```diff
@@ -694,6 +705,8 @@ void ProducerImpl::closeAsync(CloseCallback originalCallback) {
         }
     };
 
+    Lock lock(mutex_);
```

Comment:
Could you use another mutex to synchronize `closeAsync` and `handleCreatedProducer`? The `mutex_` is already used in other places and it's not clear why the mutex is required here from the code.

Reply:
For example, if someone wanted to remove the unnecessary mutex uses in future, the mutex here would be likely to be removed because the state is atomic, and the following code like cancelling the timers and failing the pending messages doesn't require a lock.

Reply:
It makes sense to me.